core-c-bridge/src/client.rs

```diff
@@ -0,0 +1,677 @@
+#[unsafe(no_mangle)]
+pub extern "C" fn temporal_core_client_rpc_call(
+    client: *mut Client,
+    options: *const RpcCallOptions,
+    user_data: *mut libc::c_void,
+    callback: ClientRpcCallCallback,
+) {
+    let client = unsafe { &*client };
+    let options = unsafe { &*options };
+    let cancel_token = unsafe { options.cancellation_token.as_ref() }.map(|v| v.token.clone());
+    let user_data = UserDataHandle(user_data);
+    client.runtime.core.tokio_handle().spawn(async move {
+        let res = match options.service {
+            RpcService::Workflow => {
+                service_call!(call_workflow_service, client, options, cancel_token)
+            }
+            RpcService::Cloud => {
+                service_call!(call_cloud_service, client, options, cancel_token)
+            }
+            RpcService::Operator => {
+                service_call!(call_operator_service, client, options, cancel_token)
+            }
+            RpcService::Test => service_call!(call_test_service, client, options, cancel_token),
+            RpcService::Health => service_call!(call_health_service, client, options, cancel_token),
+        };
+        let (success, status_code, failure_message, failure_details) = match res {
+            Ok(b) => (
+                ByteArray::from_vec(b).into_raw(),
+                0,
+                std::ptr::null_mut(),
+                std::ptr::null_mut(),
+            ),
+            Err(err) => match err.downcast::<tonic::Status>() {
+                Ok(status) => (
+                    std::ptr::null_mut(),
+                    status.code() as u32,
+                    ByteArray::from_utf8(status.message().to_string()).into_raw(),
+                    ByteArray::from_vec(status.details().to_owned()).into_raw(),
+                ),
+                Err(err) => (
+                    std::ptr::null_mut(),
+                    0,
+                    ByteArray::from_utf8(format!("{err}")).into_raw(),
+                    std::ptr::null_mut(),
+                ),
+            },
+        };
+        unsafe {
+            callback(
+                user_data.into(),
+                success,
+                status_code,
+                failure_message,
+                failure_details,
+            );
+        }
+    });
+}
+
+macro_rules! rpc_call {
+    ($client:ident, $call:ident, $call_name:ident) => {
+        if $call.retry {
+            rpc_resp($client.$call_name(rpc_req($call)?).await)
+        } else {
+            rpc_resp($client.into_inner().$call_name(rpc_req($call)?).await)
+        }
+    };
+}
+
+macro_rules! rpc_call_on_trait {
+    ($client:ident, $call:ident, $trait:tt, $call_name:ident) => {
+        if $call.retry {
+            rpc_resp($trait::$call_name(&mut $client, rpc_req($call)?).await)
+        } else {
+            rpc_resp($trait::$call_name(&mut $client.into_inner(), rpc_req($call)?).await)
+        }
+    };
+}
+
+async fn call_workflow_service(
```

Comment:
I think we should either change this to be code generated (w/ CI check confirming code gen is accurate), or something else to confirm that every string is represented here. In .NET we do have a test IIRC to confirm every gRPC call is available. In Ruby we ended up writing a code gen to generate these big match statements (though if the match statement gets too big it can have stack problems in debug build).

If we'd like though, we can defer this to another PR. But we do need something at some point to confirm this doesn't get out of sync with the proto RPC names.

Reply:
Thanks! This is good enough for me. I do think we may want code gen one day, but so long as the test is there to make sure we have to add them as they come up, sounds good.